[gpx] Specify xmlns for gpx color export

cyber-toad opened the pull request:
Fixes #8332, #9969 
Initial PR for discussions and testing, not for merge yet

cyber-toad committed:
[gpx] Add xml schema to color in gpx export

Signed-off-by: cyber-toad <the.cyber.toad@proton.me>

cyber-toad commented:
Hi, @RSR3455. Could you please check with testing for #9969 and check this file (remove .txt suffix after download) - is it visible in Garmin Basemap app and does route has green color?
Thanks.
[export_test.gpx.txt](https://github.com/user-attachments/files/18461782/export_test.gpx.txt)

cyber-toad commented:
@0709wiwiwi I tried to implement schema that you suggested - could you please have a look on file attached above - it is a test file stored with new schema. Is it in sync with what you suggested, is it valid in validators that you used (I tried one online tool and it was fine, not sure how strict it is) and is it displayed with proper color in apps that you checked?

cyber-toad commented:
Verified with gpx studio - color is supported
![image](https://github.com/user-attachments/assets/42a19a73-2f09-4be7-8970-208c69494732)

RSR3455 commented:
Basecamp gave an error message when I attempted to open the file after removing the .txt suffix, so I was unable to successfully open the file into Basecamp.

FYI, Basecamp is a free application from Garmin and can be downloaded using the following link: https://www.garmin.com/en-US/software/basecamp/.  Both Mac and Windows versions are available for download.


Robert Rearden
Fort Worth, Texas
USA
________________________________________


On Jan 17, 2025, at 5:40 PM, cyber-toad ***@***.***> wrote:


Hi, @RSR3455 <https://github.com/RSR3455>. Could you please help with testing for #9969 <https://github.com/organicmaps/organicmaps/issues/9969> and check this file (remove .txt suffix after download) - is it visible in Garmin Basemap app and does route has green color?
Thanks.
export_test.gpx.txt <https://github.com/user-attachments/files/18461782/export_test.gpx.txt>
—
Reply to this email directly, view it on GitHub <https://github.com/organicmaps/organicmaps/pull/10089#issuecomment-2599377529>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/BOBYBHNUDF4WXR3HCRLDBNL2LGIH5AVCNFSM6AAAAABVM4SS5GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZGM3TONJSHE>.
You are receiving this because you were mentioned.

RSR3455 commented:
I should have added to my earlier email that I am a GPS user with limited knowledge of GPX schema structure. I am glad to be a tester, but you may prefer to download the Basecamp application to your test sandbox.


Robert Rearden
Fort Worth, Texas
USA
________________________________________


On Jan 17, 2025, at 8:11 PM, Robert Rearden ***@***.***> wrote:

Basecamp gave an error message when I attempted to open the file after removing the .txt suffix, so I was unable to successfully open the file into Basecamp.

FYI, Basecamp is a free application from Garmin and can be downloaded using the following link: https://www.garmin.com/en-US/software/basecamp/.  Both Mac and Windows versions are available for download.


Robert Rearden
Fort Worth, Texas
USA
________________________________________


On Jan 17, 2025, at 5:40 PM, cyber-toad ***@***.***> wrote:


Hi, @RSR3455 <https://github.com/RSR3455>. Could you please help with testing for #9969 <https://github.com/organicmaps/organicmaps/issues/9969> and check this file (remove .txt suffix after download) - is it visible in Garmin Basemap app and does route has green color?
Thanks.
export_test.gpx.txt <https://github.com/user-attachments/files/18461782/export_test.gpx.txt>
—
Reply to this email directly, view it on GitHub <https://github.com/organicmaps/organicmaps/pull/10089#issuecomment-2599377529>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/BOBYBHNUDF4WXR3HCRLDBNL2LGIH5AVCNFSM6AAAAABVM4SS5GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZGM3TONJSHE>.
You are receiving this because you were mentioned.

cyber-toad commented:
Thanks, @RSR3455. I was able to install this app and check what is the issue with my latest file - some XML validity things should be fixed. But finally I was able to generate a valid GPX and it is available in Garmin Map with color applied. So once this PR is released you will be able to export compatible tracks from OM.
![1](https://github.com/user-attachments/assets/93c306db-2f32-4712-8f8a-4938778f37a4)

cyber-toad commented:
@0709wiwiwi could you please check attached file with your validators/apps?
I tried to apply schemas that you suggested and was able to have a valid color in some apps.
[export_color.gpx.txt](https://github.com/user-attachments/files/18464086/export_color.gpx.txt)

0709wiwiwi commented:
The file is not valid due to the lack of <gpx version="1.1”
  creator="cyber-toad”
Once in production optimize this as follows: creator="Organic Maps version .........” 
Garmin is merciless on this point. Other apps are usually a little more tolerant ;-)

0709wiwiwi commented:
One more note.  Speech confusion in diverse gps apps often arises as follows.

In quite a few apps, the gpx waypoints are referred to as being routepoints.
So they are not that in the first place, they just indicate an individual location.
But you can promote them into route(way)points that the routing engine will then work with. 
Route(way) points are then “sentenced” to be “snapped” to a nearby street by the router.
Associated track points are so produced by each extra via or shaping added routepoint.
Routepoints with which you draw manual tracksections are NOT handled by the router engine. 
The created "beeline" section trackpoints copy the original routepoint position unchanged.

More here: https://github.com/devemux86/cruiser/discussions/250

If you asks AI some info about this item it will give you false answers. 
AI is also sometimes confused and definitely is still learning.

cyber-toad commented:
> The file is not valid due to the lack of <gpx version="1.1” creator="cyber-toad” Once in production optimize this as follows: creator="Organic Maps version .........” Garmin is merciless on this point. Other apps are usually a little more tolerant ;-)

Yes, I initially uploaded a file without valid gpx header, but I removed this comment, but probably before you downloaded the file. I fixed it and not it contains a file that passes Notepad++ checks.

cyber-toad commented:
@biodranik @rtsisyk could you please review c++ changes. 
I worry a bit about assignment of color string here:

```
      if (distance < min) {
        min = distance;
        color = garminName;
      }
```
will it generate a copy so better to try to switch to string_view and move map outside of this function to member class?

RSR3455 commented:
Thank you for your efforts to resolve this issue.

biodranik commented on kml/kml_tests/gpx_tests.cpp:
```diff
@@ -40,6 +40,22 @@ void importExportCompare(char const * testFile)
 
+void importExportCompare(char const * sourceFile, char const * destinationFile)
+{
+  auto const sourceFileName = GetPlatform().TestsDataPathForFile(sourceFile);
+  std::string sourceFileText;
+  FileReader(sourceFileName).ReadAsString(sourceFileText);
+  kml::FileData const dataFromFile = loadGpxFromFile(sourceFile);
+  std::string resultBuffer;
+  MemWriter<decltype(resultBuffer)> sink(resultBuffer);
+  kml::gpx::SerializerGpx ser(dataFromFile);
```
nit: copy paste.

cyber-toad replied:
Fixed

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -34,8 +34,14 @@ std::string_view constexpr kCmt = "cmt";
 std::string_view constexpr kGpxHeader =
-    "<?xml version=\"1.0\"?>\n"
-    "<gpx xmlns=\"http://www.topografix.com/GPX/1/1\" version=\"1.1\">\n";
+  "<?xml version=\"1.0\"?>\n"
```
nit: Using raw string literals can be more readable: https://en.cppreference.com/w/cpp/language/string_literal

cyber-toad replied:
Fixed

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -133,7 +139,12 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
 {
-  auto const colorBytes = FromHex(value);
+  std::string_view colorStr = value;
+  if (colorStr.at(0) == '#')
```
1. at() throws an exception if index is invalid, so it is slower than .front() or colorStr[0].
2. Is value always non-empty here?

cyber-toad replied:
We could have some invalid file with empty color value, so I reworked it a bit to add a check

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -133,7 +139,12 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
 
 void GpxParser::ParseColor(std::string const & value)
```
```suggestion
void GpxParser::ParseColor(std::string_view value)
```

cyber-toad replied:
Fixed

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -133,7 +139,12 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
+  if (colorStr.at(0) == '#')
+    colorStr = colorStr.substr(1, colorStr.size() - 1);
+  if (colorStr.length() == 8)
+    colorStr = colorStr.substr(2, colorStr.length() - 1);
```
Consistency: either always use .size() or always use .length()

cyber-toad replied:
Fixed

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -133,7 +139,12 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
-  auto const colorBytes = FromHex(value);
+  std::string_view colorStr = value;
+  if (colorStr.at(0) == '#')
+    colorStr = colorStr.substr(1, colorStr.size() - 1);
```
Will .remove_prefix(1) work?

cyber-toad replied:
Fixed

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -133,7 +139,12 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
+  std::string_view colorStr = value;
+  if (colorStr.at(0) == '#')
+    colorStr = colorStr.substr(1, colorStr.size() - 1);
+  if (colorStr.length() == 8)
```
What is this case?

cyber-toad replied:
We have something like <xsi:gpx><color>#FF00FF00</color></xsi:gpx> in gpx. It is handled by default parser (not Osmand/etc) and earlier it was not ready for such representation

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -133,7 +139,12 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
+  if (colorStr.length() == 8)
+    colorStr = colorStr.substr(2, colorStr.length() - 1);
+  auto const colorBytes = FromHex(colorStr);
   if (colorBytes.size() != 3)
```
Can color be transparent?

cyber-toad replied:
In theory it could be in gpx.
But later we have```

  if (colorBytes.size() != 3)
  {
    LOG(LWARNING, ("Invalid color value", value));
```
so it will reject such colors, so when I saw it I added additional code to avoid issue if we have 4 components

biodranik replied:
Can drape engine render semitransparent tracks? If yes, then it may be good to support additional transparent byte. Otherwise it can be ignored/forced to always opaque.

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +494,66 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+         << NumToHex(static_cast<uint8_t>((rgba >> 8) & 0xFF));
+}
+
+std::tuple<uint8_t, uint8_t, uint8_t> ExtractRGB(std::uint32_t color) {
```
```suggestion
std::tuple<uint8_t, uint8_t, uint8_t> ExtractRGB(uint32_t color)
{
```

cyber-toad replied:
Fixed

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +494,66 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+  return {(color >> 24) & 0xFF, (color >> 16) & 0xFF, (color >> 8) & 0xFF};
+}
+
+int ColorDistance(std::uint32_t color1, std::uint32_t color2) {
```
```suggestion
int ColorDistance(uint32_t color1, uint32_t color2)
{
```

cyber-toad replied:
Fixed

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +494,66 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+}
+
+int ColorDistance(std::uint32_t color1, std::uint32_t color2) {
+  auto [r1, g1, b1] = ExtractRGB(color1);
+  auto [r2, g2, b2] = ExtractRGB(color2);
```
nit
```suggestion
  auto const [r1, g1, b1] = ExtractRGB(color1);
  auto const [r2, g2, b2] = ExtractRGB(color2);
```

cyber-toad replied:
Fixed

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +494,66 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+  };
+  auto const it = kHexToGarmin.find(rgba);
+  if (it != kHexToGarmin.end())
+  {
+    return it->second;
+  }
```
```suggestion
    return it->second;
```

cyber-toad replied:
Fixed

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +494,66 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+  {
+    return it->second;
+  }
+  const auto & first = kHexToGarmin.begin();
```
```suggestion
  const auto & [min, color] = *kHexToGarmin.begin();
```

cyber-toad replied:
This makes color const, but we assign it inside loop later

biodranik replied:
Make it non-const then )

cyber-toad replied:
I slightly reworked it and now deconstructor is used as suggested.

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +494,66 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+  return (r1 - r2) * (r1 - r2) + (g1 - g2) * (g1 - g2) + (b1 - b2) * (b1 - b2);
+}
+
+std::string MapGarminColor(uint32_t rgba)
```
Is there a test for this method?

cyber-toad replied:
It is covered in Gpx_ImportExport_Test where color from hashtable is used and in Gpx_ColorMapExport_Test where 0x443344 that is not in the table is mapped to closest one (DarkGray)

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +494,66 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+
+std::string MapGarminColor(uint32_t rgba)
+{
+  static std::unordered_map<std::uint32_t, std::string> const kHexToGarmin = {
```
```suggestion
  static std::unordered_map<std::uint32_t, std::string_view> const kHexToGarmin = {
```

cyber-toad replied:
Should I make function to return std::string_view for this change or create std::string before return? If  use string_view for color, I got a warning in IDE that address of local variable color escaped the function, I'm not confident in my cpp knowledge to get is it an issue :) (I remember we had some problems with memory in gpx parser when it was reworked to string_view).

cyber-toad replied:
![image](https://github.com/user-attachments/assets/7d5397d0-88bd-44b3-81e2-4f09b32fd5ec)

Here is a clion warning for return for the version of code that you suggest here: https://github.com/organicmaps/organicmaps/pull/10089#discussion_r1921647356
But godbolt doesn't warn about it: https://godbolt.org/z/ah5q9e339

biodranik replied:
Looks like clang-tidy (or whatever tool is used by clion) doesn't detect that constexpr kRGBAToGarmin is also a static variable in this case. You may try to add `static` there explicitly. Btw, it generates a bit smaller code for x86 gcc.

cyber-toad replied:
I tried to replace with `auto static constexpr kRGBAToGarmin` and warning is the same. based on this hhttps://youtrack.jetbrains.com/issues?q=%20%22the%20address%20of%20the%20local%20variable%20may%20escape%20the%20function%22%20 analysis is done by clion itself and they are fixing it from time to time. So if you don't see issues with it we can try to run test suite and see how it goes.

biodranik replied:
The code should work, but it is always better to write unit tests and run them on ci with different compilers.

cyber-toad replied:
Done

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +494,66 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+    return it->second;
+  }
+  const auto & first = kHexToGarmin.begin();
+  u_int32_t min = first->first;
```
Will the code below work properly if this min is some color from garmin table?

cyber-toad replied:
yes, there is a typo there, should be `min = ColorDistance(first->first, rgba); ` I fixed it

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -137,13 +134,18 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
-void GpxParser::ParseColor(std::string const & value)
+void GpxParser::ParseColor(std::string_view value)
 {
   std::string_view colorStr = value;
```
Is this line needed? Why `value` (better rename it to `colorStr`) can not be used directly?

cyber-toad replied:
Some leftover after multiple changes, renamed it in method signature

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -502,13 +504,15 @@ void SaveColorToARGB(Writer & writer, uint32_t rgba)
 }
 
-std::tuple<uint8_t, uint8_t, uint8_t> ExtractRGB(std::uint32_t color) {
+std::tuple<uint8_t, uint8_t, uint8_t> ExtractRGB(uint32_t color)
```
Internal .cpp helper functions should be either static or in anonymous `namespace` to avoid being exported.

cyber-toad replied:
I think they are already in `namespace {` - line 480?

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -502,13 +504,15 @@ void SaveColorToARGB(Writer & writer, uint32_t rgba)
-int ColorDistance(std::uint32_t color1, std::uint32_t color2) {
-  auto [r1, g1, b1] = ExtractRGB(color1);
-  auto [r2, g2, b2] = ExtractRGB(color2);
+int ColorDistance(uint32_t color1, uint32_t color2)
```
ditto

cyber-toad replied:
I think they are already in `namespace {` - line 480?

biodranik replied:
Right!

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -535,11 +539,9 @@ std::string MapGarminColor(uint32_t rgba)
-  }
   const auto & first = kHexToGarmin.begin();
-  u_int32_t min = first->first;
+  u_int32_t min = ColorDistance(first->first, rgba);
```
What is u_int32_t?

cyber-toad replied:
I made a typo, but cpp has type for it also :)  Fixed

cyber-toad commented:
> Can drape engine render semitransparent tracks? If yes, then it may be good to support additional transparent byte. Otherwise it can be ignored/forced to always opaque.

By some reason can't reply to initial comment.
I tried to set 128 to alpha instead of 255 and track was displayed as earlier, so probably it is not supported yet.

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +495,63 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+    return it->second;
+  auto [hex, closestColor] = *kHexToGarmin.begin();
+  uint32_t minDistance = ColorDistance(hex, rgba);
+  for (const auto & [garminColor, garminName] : kHexToGarmin)
```
As a bonus, you can move this constexpr array into an anonymous namespace, and reuse it in ParseGarminColor method.

Maybe ParseOsmandColor can also be improved using string_view.

These improvements can be done in a separate PR

cyber-toad commented on kml/kml_tests/gpx_tests.cpp:
```diff
@@ -11,7 +11,7 @@
 namespace gpx_tests
 {
-static kml::FileData loadGpxFromString(std::string_view content) {
+static kml::FileData LoadGpxFromString(std::string_view content) {
```
fixed formatting

cyber-toad commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +495,63 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+    {0xff00ffff, "Magenta"},
+    {0x00ffffff, "Cyan"},
+    {0xffffffff, "White"},
+    {0xff0000ff, "Transparent"}
```
yes, this one is not required. fixed and removed

cyber-toad commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +495,63 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+  return {(color >> 24) & 0xFF, (color >> 16) & 0xFF, (color >> 8) & 0xFF};
+}
+
+uint32_t ColorDistance(uint32_t color1, uint32_t color2)
```
Got it, in this case I think the more important is to have 
`std::tuple<int, int, int> ExtractRGB(uint32_t color)` return tuple of ints because we subtract outputs of this function?  Result of color distance is always a positive value, so it should be safe to store and compare it as uint32_t? But I agree, better to avoid unsigned when we can.

cyber-toad replied:
We don't substract result of ColorDistance, but just compare it and it is always some posistive value, so should work with uint32_t also? But I agree that better to avoid unsigned when we can.

cyber-toad commented on kml/serdes_gpx.cpp:
```diff
@@ -483,6 +495,63 @@ void SaveColorToRGB(Writer & writer, uint32_t rgba)
+         << NumToHex(static_cast<uint8_t>((rgba >> 8) & 0xFF));
+}
+
+std::tuple<uint8_t, uint8_t, uint8_t> ExtractRGB(uint32_t color)
```
Updated to signed

cyber-toad commented:
```
    Event if drape doesn't draw such colors as transparent now, maybe we can support it later.
    If the track/gpx/kml already has some transparency defined, would it be better to store it for the future and not lose on import/export?
```
transparency is not always parsed as a part of color (http://www.topografix.com/gpx/gpx_style/0/2/#type_opacityType), so these formats need to be supported in parser and writer. but maybe we can cover some cases here.

cyber-toad commented on kml/serdes_gpx.cpp:
```diff
@@ -131,12 +134,21 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
+  if (colorStr.front() == '#')
+    colorStr.remove_prefix(1);
+  if (colorStr.size() == 8)
+    colorStr = colorStr.substr(2, colorStr.size() - 1);
```
yes, I made a mistake thinking that parameters are [from, to], so the code is wrong but worked due to termination symbol. I replaced it with remove_prefix

cyber-toad commented on kml/serdes_gpx.cpp:
```diff
@@ -131,12 +134,21 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
 }
 
-void GpxParser::ParseColor(std::string const & value)
+void GpxParser::ParseColor(std::string_view colorStr)
```
I changed it and added tests. But I can't test all cases because there is an assert in impl::HexDigitToRaw that terminates execution for tests data like #MMMMMM that has valid length but wrong content. In this case should I disable assertions for these tests somehow? What is the prefered way to do it - I googled #define NDEBUG but looks like OM uses some custom assertions in assert.hpp.

cyber-toad commented:
> If the track/gpx/kml already has some transparency defined, would it be better to store it for the future and not lose on import/export?

Added store of opacity if it is encoded as a part of color

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -131,42 +134,44 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
+
+void GpxParser::ParseColor(std::string_view colorStr)
+{
+  if (const auto parsed = ParseColorFromHexString(colorStr); parsed.has_value())
```
nit: (less letters usually means a bit more readable code)
```suggestion
  if (const auto parsed = ParseColorFromHexString(colorStr); parsed)
```

cyber-toad replied:
Fixed

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -131,42 +134,44 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
+// https://osmand.net/docs/technical/osmand-file-formats/osmand-gpx/. Supported colors: #AARRGGBB/#RRGGBB/AARRGGBB/RRGGBB
+void GpxParser::ParseOsmandColor(std::string const & value)
+{
+  std::optional<uint32_t> parsed = ParseColorFromHexString(value);
+  if (!parsed.has_value())
```
nit:
```suggestion
  auto const color = ParseColorFromHexString(value);
  if (!color)
```

Should a warning be printed if color fails to parse?

cyber-toad replied:
There is some logging in GpxParser::ParseColorFromHexString for unexpected values, do you want to add some more logging?

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -131,42 +134,44 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
+  std::optional<uint32_t> parsed = ParseColorFromHexString(value);
+  if (!parsed.has_value())
+    return;
+  uint32_t color = parsed.value();
```
```suggestion
```

cyber-toad replied:
Fixed

biodranik commented on kml/serdes_gpx.cpp:
```diff
@@ -131,42 +134,44 @@ std::string const & GpxParser::GetTagFromEnd(size_t n) const
+  uint32_t color = parsed.value();
   if (m_tags.size() > 2 && GetTagFromEnd(2) == gpx::kGpx)
   {
     m_globalColor = color;
```
```suggestion
    m_globalColor = *color;
```

And the same below.

cyber-toad replied:
Fixed

biodranik reviewed (approved):
Thanks! LGTM with a few nits.

> transparency is not always parsed as a part of color (http://www.topografix.com/gpx/gpx_style/0/2/#type_opacityType), so these formats need to be supported in parser and writer. but maybe we can cover some cases here.

How is this style spec related to the main gpx spec? I didn't find any reference to styles from gpx 1.1.

> RE: MapGarminColor in unit tests

There are several options for how to do that with their pros and cons. The simplest one is to move the function outside of the anonymous namespace (leave it in kml::gpx), and not make it static in the cpp file. Then you can reference it in another cpp by either including a header where it is declared, or (simpler) declaring it right before using it in the cpp unit test file:
```c++
namespace kml::gpx
{
std::string_view MapGarminColor(uint32_t rgba);
}  // namespace kml::gpx 
```

A perfect approach would use a test-only time define (e.g. derived from cmake's BUILD_TESTING), and use anonymous namespace (or `static` keyword) if production binary (or dynamic library in case of Android) is built, and expose the function if libraries are built for unit tests.

cyber-toad commented on kml/kml_tests/gpx_tests.cpp:
```diff
@@ -340,9 +366,25 @@ d5
 
+UNIT_TEST(ParseFromString)
+{
+  // string hex sequence #AARRGGBB, uint32 sequence RGBA
```
Fixed

cyber-toad commented:
> How is this style spec related to the main gpx spec? I didn't find any reference to styles from gpx 1.1.

Yes, I mean that there are different ways how different applications exports opacity. For this ticket I added a case when it is a part of the color, but there are some schemas when it is stored separately.

cyber-toad commented:
> There are several options for how to do that with their pros and cons. The simplest one is to move the function outside of the anonymous namespace (leave it in kml::gpx), and not make it static in the cpp file.

Thanks, I moved function and added a test for it, please have a look.
